### PR TITLE
mon: do not alloc monmap on heap in Elector::dispatch()

### PR DESCRIPTION
**Pull request by songbaisen**
mon: elector dispatch do not alloc memory on monmap

Signed-off-by: song baisen song.baisen@zte.com.cn

### CODE REVIEW

File: src/mon/Elector.cc [hunk 1]
```diff
@@ -453,7 +452,6 @@ void Elector::dispatch(MonOpRequestRef op)
 		<< " < my epoch " << mon->monmap->epoch 
 		<< dendl;
       } 
-      delete peermap;
```
Comment (branch-predictor) [inline src/mon/Elector.cc]: Do we really need to construct MonMap instance by hand?

**Comment by songbaisen** [comment]
@branch-predictor  Good eyes man..For performance here is really does not need to new memory every time.

File: src/mon/Elector.h [hunk 2]
```diff
@@ -44,7 +45,11 @@ class Elector {
    * The Monitor instance associated with this class.
    */
   Monitor *mon;
-
+  /**
+   * The instance of the peer monitor map
+   *
+   */ 
+  MonMap peermap;
```
Comment (tchaikov) [inline src/mon/Elector.h]: why would you want to keep the peermap around as a member variable?

Comment (songbaisen) [inline src/mon/Elector.h]: @tchaikov  consider the  branch-predictor  idea.do not need to  new and free every time.

File: src/mon/Elector.cc [hunk 3]
```diff
@@ -430,10 +430,9 @@ void Elector::dispatch(MonOpRequestRef op)
 	return;
       }
 
-      MonMap *peermap = new MonMap;
-      peermap->decode(em->monmap_bl);
-      if (peermap->epoch > mon->monmap->epoch) {
-	dout(0) << em->get_source_inst() << " has newer monmap epoch " << peermap->epoch
+      peermap.decode(em->monmap_bl);
```
Comment (tchaikov) [inline src/mon/Elector.cc]: could just use a local variable? guess that's what @branch-predictor means.

**Comment by songbaisen** [comment]
@tchaikov  Thank you sir! Got it.

**Comment by yangdongsheng** [comment]
@songbaisen Sorry, but I would ask, what's the benefit from this change? what I can see here is just moving the peermap location from heap to stack, what's the difference? Could you give a more explanation? thanx

**Comment by songbaisen** [comment]
@yangdongsheng  Hello.Nice to meet you.Thank you for your reply.Here is not very a lot difference.Just for the safe in future.If any body change the code before delete,we may have a risk of memory leak.For prevent that happen change it.

**Comment by branch-predictor** [comment]
@tchaikov @songbaisen yep, I meant having an MonMap instance as local variable. This way we're leak-free *and* we're one heap (de)allocation less.

**Comment by athanatos** [comment]
Passed tests: sjust@teuthology:/a/samuelj-2016-05-16_20:02:21-rados-wip-sam-testing-distro-basic-smithi

@tchaikov Does this look ok?

**Pull request merged**